src/vm_interrupts.rs

```diff
@@ -136,9 +137,96 @@ impl VmCpuExtInterrupts {
         // Finish restoring state from the SW interrupt file.
         let imsic = Imsic::get();
         imsic
-            .restore_guest_file_finish(interrupt_file, &mut self.sw_file)
+            .restore_guest_file(interrupt_file, &mut self.sw_file)
+            .map_err(Error::BindingImsic)?;
+        self.bind_status = BindStatus::Bound(cpu, interrupt_file);
+        Ok(interrupt_file)
+    }
+
+    /// Prepares to rebind this vCPU to `interrupt_file` on the current physical CPU.
+    pub fn rebind_imsic_prepare(&mut self, interrupt_file: ImsicFileId) -> Result<()> {
+        // The vCPU must be bound to an other IMSIC file.
+        let (cpu, old_interrupt_file) = match self.bind_status {
+            BindStatus::Bound(cpu, old_interrupt_file) => (cpu, old_interrupt_file),
+            _ => return Err(Error::WrongBindStatus),
+        };
+
+        // Initialize the guest IMSIC file we're getting bound to.
+        Imsic::get()
+            .clear_guest_file(interrupt_file)
+            .map_err(Error::BindingImsic)?;
+        self.bind_status = BindStatus::Rebinding(
+            cpu,
+            old_interrupt_file,
+            PerCpu::this_cpu().cpu_id(),
+            interrupt_file,
+        );
+        Ok(())
+    }
+
+    /// Copies the guest interrupt file over to sw file.
+    pub fn rebind_imsic_clone(&mut self) -> Result<()> {
+        // Make sure there's a rebind operation in progress and we are on old CPU right now.
+        let (cpu, interrupt_file, new_cpu, new_file) = match self.bind_status {
+            BindStatus::Rebinding(cpu, interrupt_file, new_cpu, new_file) => {
+                (cpu, interrupt_file, new_cpu, new_file)
+            }
+            _ => return Err(Error::WrongBindStatus),
+        };
+
+        if cpu != PerCpu::this_cpu().cpu_id() {
+            return Err(Error::WrongPhysicalCpu);
+        }
+
+        // Copy and then clear the guest interrupt file.
+        let imsic = Imsic::get();
+        imsic
+            .save_guest_file_prepare(interrupt_file, &mut self.sw_file)
             .map_err(Error::BindingImsic)?;
+        imsic
+            .save_guest_file_finish(interrupt_file, &mut self.sw_file)
+            .map_err(Error::RebindingImsic)?;
+        self.bind_status = BindStatus::Cloned(cpu, interrupt_file, new_cpu, new_file);
+        Ok(())
+    }
+
+    /// Get the previous guest interrupt file's location.
+    pub fn prev_imsic_location(&self) -> Result<ImsicLocation> {
+        // Make sure we are in the rebinding state.
+        let (cpu, interrupt_file) = match self.bind_status {
+            BindStatus::Rebinding(cpu, interrupt_file, _, _) => (cpu, interrupt_file),
+            BindStatus::Cloned(cpu, interrupt_file, _, _) => (cpu, interrupt_file),
```

Comment:
See comment above: `Rebinding` should be the only state where we care about the previous location.